File: python/rotate_string.py
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        """time complexity: O(n)"""
        if len(s) == len(goal) and len(s) != 0 and len(goal) != 0:
            return self.is_substring(s + s, goal)
        return False

    def is_substring(self, s: str, t: str) -> bool:
        """time complexity: O(n)"""
        return s.find(t) != -1


def test_rotateString_case_1():
    # arrange
    s: str = "abcde"
    goal: str = "cdeab"
    expected: bool = True

    # act
    actual = Solution().rotateString(s, goal)

    # assert
    assert expected == actual


def test_rotateString_case_2():
    # arrange
    s: str = "abcde"
    goal: str = "abced"
    expected: bool = False

    # act
    actual = Solution().rotateString(s, goal)

    # assert
    assert expected == actual

File: rust/src/code/rotate_string.rs
struct Solution;

impl Solution {
    // time complexity: O(n)
    pub fn rotate_string(s: String, goal: String) -> bool {
        if s.len() == goal.len() && s.len() != 0 {
            let s: String = s.clone() + &s;
            if s.contains(&goal) {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_rotate_string_case_1() {
        // arrange
        let s: String = "abcde".to_string();
        let goal: String = "cdeab".to_string();
        let expected: bool = true;

        // act
        let actual: bool = Solution::rotate_string(s, goal);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_rotate_string_case_2() {
        // arrange
        let s: String = "abcde".to_string();
        let goal: String = "abced".to_string();
        let expected: bool = false;

        // act
        let actual: bool = Solution::rotate_string(s, goal);

        // assert
        assert_eq!(expected, actual);
    }
}
